### ISSUE TITLE
Deleting an immutable TAM object should fail instead of recursively deleting

### ISSUE THREAD

**Issue by nchaimov** [ParaToolsInc/taucmdr#364]
Once a TAM object has been used in an Experiment that has at least one Trial, the TAM objects making up the experiment become immutable. However, it is still possible to delete them, which has the surprising result of deleting all Experiments that make use of the object, which in turn deletes all of that Experiment's Trials.

For example, I initialize a new project in `examples/mm`, switch to an experiment using the `baseline` measurement, and run a Trial: 
 
```
nchaimov@delphi /home/nchaimov/taucmdr/examples/mm $ tau init
[...]
nchaimov@delphi /home/nchaimov/taucmdr/examples/mm $ tau select baseline
[TAU] Created a new experiment 'delphi-mm-baseline'
[TAU] Selected experiment 'delphi-mm-baseline'.
nchaimov@delphi /home/nchaimov/taucmdr/examples/mm $ make
tau gcc   -c matmult.c -o matmult.o
tau gcc   -c matmult_initialize.c -o matmult_initialize.o
tau gcc   matmult.o matmult_initialize.o -o matmult.exe
nchaimov@delphi /home/nchaimov/taucmdr/examples/mm $ tau ./matmult.exe
[...]
[TAU] Trial 0 produced 1 profile files.
[TAU]
[TAU] == END delphi-mm-baseline at 2020-05-11 22:02:00.359232 =============================================================
[TAU]
[TAU] Experiment: delphi-mm-baseline
[TAU] Command: tau_baseline ./matmult.exe
[TAU] Current working directory: /home/nchaimov/taucmdr/examples/mm
[TAU] Data size: 2.5KiB bytes
[TAU] Elapsed seconds: 2.62067985535
```
 
The dashboard now shows the new Trial:
 
```
nchaimov@delphi /home/nchaimov/taucmdr/examples/mm $ tau dash

== Project Configuration (/home/nchaimov/taucmdr/examples/mm/.tau/project.json) ===============================================================

+------+---------+--------------+--------------------------------------------------------------+---------------+
| Name | Targets | Applications |                         Measurements                         | # Experiments |
+======+=========+==============+==============================================================+===============+
|   mm | delphi  |      mm      | baseline, sample, profile, source-inst, compiler-inst, trace |       2       |
+------+---------+--------------+--------------------------------------------------------------+---------------+

== Targets in project 'mm' ====================================================================================================================

+--------+---------+-----------+----------------+---------------+-----------------+
|  Name  | Host OS | Host Arch | Host Compilers | MPI Compilers | SHMEM Compilers |
+========+=========+===========+================+===============+=================+
| delphi |  Linux  |  x86_64   |      GNU       |    System     |    OpenSHMEM    |
+--------+---------+-----------+----------------+---------------+-----------------+

== Applications in project 'mm' ===============================================================================================================

+------+---------+--------+----------+-----+-----+------+--------+-------+-----+
| Name | Linkage | OpenMP | Pthreads | TBB | MPI | CUDA | OpenCL | SHMEM | MPC |
+======+=========+========+==========+=====+=====+======+========+=======+=====+
|   mm | dynamic |   No   |    No    | No  | No  |  No  |   No   |  No   | No  |
+------+---------+--------+----------+-----+-----+------+--------+-------+-----+

== Measurements in project 'mm' ===============================================================================================================

+---------------+---------+-------+--------+--------------+----------------+--------+------+-----+-----+-------+
|     Name      | Profile | Trace | Sample | Source Inst. | Compiler Inst. | OpenMP | CUDA | I/O | MPI | SHMEM |
+===============+=========+=======+========+==============+================+========+======+=====+=====+=======+
|      baseline |   tau   | none  |   No   |    never     |     never      | ignore |  No  | No  | No  |  No   |
+---------------+---------+-------+--------+--------------+----------------+--------+------+-----+-----+-------+
|        sample |   tau   | none  |  Yes   |    never     |     never      | ignore |  No  | No  | No  |  No   |
+---------------+---------+-------+--------+--------------+----------------+--------+------+-----+-----+-------+
|       profile |   tau   | none  |   No   |  automatic   |    fallback    | ignore |  No  | No  | No  |  No   |
+---------------+---------+-------+--------+--------------+----------------+--------+------+-----+-----+-------+
|   source-inst |   tau   | none  |   No   |  automatic   |     never      | ignore |  No  | No  | No  |  No   |
+---------------+---------+-------+--------+--------------+----------------+--------+------+-----+-----+-------+
| compiler-inst |   tau   | none  |   No   |    never     |     always     | ignore |  No  | No  | No  |  No   |
+---------------+---------+-------+--------+--------------+----------------+--------+------+-----+-----+-------+
|         trace |  none   | otf2  |   No   |  automatic   |    fallback    | ignore |  No  | No  | No  |  No   |
+---------------+---------+-------+--------+--------------+----------------+--------+------+-----+-----+-------+

== Experiments in project 'mm' ================================================================================================================

+--------------------+--------+-----------+--------+-------------+-------------+---------------+-----------------------+
|        Name        | Trials | Data Size | Target | Application | Measurement | Record Output |     TAU Makefile      |
+====================+========+===========+========+=============+=============+===============+=======================+
|   delphi-mm-sample |   0    |   0.0B    | delphi |     mm      |   sample    |       0       | Makefile.tau-0ff8ef38 |
+--------------------+--------+-----------+--------+-------------+-------------+---------------+-----------------------+
| delphi-mm-baseline |   1    |  2.5KiB   | delphi |     mm      |  baseline   |       0       |          N/A          |
+--------------------+--------+-----------+--------+-------------+-------------+---------------+-----------------------+

Selected Experiment: delphi-mm-baseline
```

and it appears when listed:

```
nchaimov@delphi /home/nchaimov/taucmdr/examples/mm $ tau trial list
== Trial Configurations (/home/nchaimov/taucmdr/examples/mm/.tau/project.json) ================================================================

+--------+-----------+----------------------------+-------------+-----------+-----------------+
| Number | Data Size |          Command           | Description |  Status   | Elapsed Seconds |
+========+===========+============================+=============+===========+=================+
|   0    |  2.5KiB   | tau_baseline ./matmult.exe |     N/A     | completed |      2.621      |
+--------+-----------+----------------------------+-------------+-----------+-----------------+

Selected experiment: delphi-mm-baseline
```
 
  
If I try to **edit** the Measurement `baseline`, this fails as expected:
 
```
nchaimov@delphi /home/nchaimov/taucmdr/examples/mm $ tau meas edit baseline --baseline False
[TAU] XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX
[TAU]
[TAU] CRITICAL
[TAU]
[TAU] Measurement 'baseline' cannot be modified because it is used by these experiments: delphi-mm-baseline
[TAU]
[TAU] Hints:
[TAU]   * Use `tau measurement copy` to create a modified copy of the measurement
[TAU]   * Use `tau experiment delete` to delete the experiments.
[TAU]
[TAU] TAU cannot proceed with the given inputs.
[TAU] Please check the configuration for errors or contact <support@paratools.com> for assistance.
[TAU]
[TAU] XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX
```
 
However, if I delete the Measurement `baseline`, it is deleted, as are the Experiment and the Trial!
 
```
nchaimov@delphi /home/nchaimov/taucmdr/examples/mm $ tau dash

== Project Configuration (/home/nchaimov/taucmdr/examples/mm/.tau/project.json) ===============================================================

+------+---------+--------------+----------------------------------------------------+---------------+
| Name | Targets | Applications |                    Measurements                    | # Experiments |
+======+=========+==============+====================================================+===============+
|   mm | delphi  |      mm      | sample, profile, source-inst, compiler-inst, trace |       1       |
+------+---------+--------------+----------------------------------------------------+---------------+

== Targets in project 'mm' ====================================================================================================================

+--------+---------+-----------+----------------+---------------+-----------------+
|  Name  | Host OS | Host Arch | Host Compilers | MPI Compilers | SHMEM Compilers |
+========+=========+===========+================+===============+=================+
| delphi |  Linux  |  x86_64   |      GNU       |    System     |    OpenSHMEM    |
+--------+---------+-----------+----------------+---------------+-----------------+

== Applications in project 'mm' ===============================================================================================================

+------+---------+--------+----------+-----+-----+------+--------+-------+-----+
| Name | Linkage | OpenMP | Pthreads | TBB | MPI | CUDA | OpenCL | SHMEM | MPC |
+======+=========+========+==========+=====+=====+======+========+=======+=====+
|   mm | dynamic |   No   |    No    | No  | No  |  No  |   No   |  No   | No  |
+------+---------+--------+----------+-----+-----+------+--------+-------+-----+

== Measurements in project 'mm' ===============================================================================================================

+---------------+---------+-------+--------+--------------+----------------+--------+------+-----+-----+-------+
|     Name      | Profile | Trace | Sample | Source Inst. | Compiler Inst. | OpenMP | CUDA | I/O | MPI | SHMEM |
+===============+=========+=======+========+==============+================+========+======+=====+=====+=======+
|        sample |   tau   | none  |  Yes   |    never     |     never      | ignore |  No  | No  | No  |  No   |
+---------------+---------+-------+--------+--------------+----------------+--------+------+-----+-----+-------+
|       profile |   tau   | none  |   No   |  automatic   |    fallback    | ignore |  No  | No  | No  |  No   |
+---------------+---------+-------+--------+--------------+----------------+--------+------+-----+-----+-------+
|   source-inst |   tau   | none  |   No   |  automatic   |     never      | ignore |  No  | No  | No  |  No   |
+---------------+---------+-------+--------+--------------+----------------+--------+------+-----+-----+-------+
| compiler-inst |   tau   | none  |   No   |    never     |     always     | ignore |  No  | No  | No  |  No   |
+---------------+---------+-------+--------+--------------+----------------+--------+------+-----+-----+-------+
|         trace |  none   | otf2  |   No   |  automatic   |    fallback    | ignore |  No  | No  | No  |  No   |
+---------------+---------+-------+--------+--------------+----------------+--------+------+-----+-----+-------+

== Experiments in project 'mm' ================================================================================================================

+------------------+--------+-----------+--------+-------------+-------------+---------------+-----------------------+
|       Name       | Trials | Data Size | Target | Application | Measurement | Record Output |     TAU Makefile      |
+==================+========+===========+========+=============+=============+===============+=======================+
| delphi-mm-sample |   0    |   0.0B    | delphi |     mm      |   sample    |       0       | Makefile.tau-0ff8ef38 |
+------------------+--------+-----------+--------+-------------+-------------+---------------+-----------------------+

No selected experiment: Use `tau select` to create or select an experiment.
```
 
The Trial data is actually deleted, it is not merely omitted from the dashboard:
 
```
nchaimov@delphi /home/nchaimov/taucmdr/examples/mm $ ls .tau/mm -la
total 0
drwxr-xr-x. 3 nchaimov cis_stud 29 May 11 15:05 .
drwxr-xr-x. 4 nchaimov cis_stud 73 May 11 14:59 ..
drwxr-xr-x. 2 nchaimov cis_stud  6 May 11 14:59 delphi-mm-sample
```
 
This behavior should be changed to prevent deletion of immutable TAM objects. Recursive deletion could perhaps be allowed if a flag to specifically enable it is passed.

**Comment by zbeekman** [comment]
> Recursive deletion could perhaps be allowed if a flag to specifically enable it is passed.

We should consider adding this functionality at some point.

Sam's PR looks like it will solve this issue and looks great in general! I'm going to verify the behavior and merge it then merge in python3 support on top of it.

**Comment by nchaimov** [comment]
Fixed in #367
